macros/path_transformations/transform_paths.sql

```diff
@@ -0,0 +1,126 @@
+/* Macro to remove complexity from models paths_to_conversion / paths_to_non_conversion. */
+
+{% macro transform_paths(model_type, source_cte) %}
+  {{ return(adapter.dispatch('transform_paths', 'snowplow_fractribution')(model_type, source_cte)) }}
+{% endmacro %}
+
+{% macro default__transform_paths(model_type, source_cte) %}
+
+  {% set allowed_path_transforms = ['exposure_path', 'first_path', 'frequency_path', 'remove_if_last_and_not_all', 'remove_if_not_all', 'unique_path'] %}
+
+  , path_transforms as (
+
+     select
+        customer_id,
+        {% if model_type == 'conversions' %}
+        conversion_tstamp,
+        revenue,
+        {% endif %}
+        {{ trim_long_path('path', var('path_lookback_steps')) }} as path,
+
+    {% if var('path_transforms').items()|length > 0 %}
+
+      {% for path_transform_name, _ in var('path_transforms').items()|reverse %}
+        {% if path_transform_name not in allowed_path_transforms %}
+          {%- do exceptions.raise_compiler_error("Snowplow Error: the path transform - '"+path_transform_name+"' - is not supported. Please refer to the Snowplow docs on tagging. Please use one of the following: exposure_path, first_path, frequency_path, remove_if_last_and_not_all, remove_if_not_all, unique_path") %}
+        {% endif %}
+        {{target.schema}}.{{path_transform_name}}(
+      {% endfor %}
+
+      transformed_path
+
+      {% for _, transform_param in var('path_transforms').items()|reverse %}
+        {% if transform_param %}, '{{transform_param}}' {% endif %}
+        )
+      {% endfor %}
+
+      as transformed_path
+
+    {% else %}
+     transformed_path
+    {% endif %}
+
+  from {{ source_cte }}
+
+  )
+
+{% endmacro %}
+
+
+{% macro databricks__transform_paths(model_type, source_cte) %}
+
+  {% set total_transformations = var('path_transforms').items()|length %}
+  -- set loop_count using namespace to define it as global variable for the loop to work
+  {% set loop_count = namespace(value=1) %}
+
+  {% for path_transform_name, transform_param in var('path_transforms').items() %}
+
+    {%- if loop_count.value == 1 %}
+      {% set previous_cte = 'arrays' %}
```

Comment:
Need to add the new `source_cte` arg here as well I think? (even though I do think it never hits this on the first itteration so maybe it's not needed at all)

Reply:
I noticed this already while fixing the above. It should hit this otherwise it would fail as these are mid script ctes.